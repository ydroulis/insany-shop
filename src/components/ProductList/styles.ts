"use client"

import styled from 'styled-components'

export const Container = styled.section`
    max-width: 1120px;

    margin: 48px auto 51px auto;

    color: #000000;

    position: relative;

    p {
        text-align: center;
        margin: 0 auto;
    }
`

export const Title = styled.h1`
    font-size: 2.5rem;
    font-weight: 700;

    margin-bottom: 47px;
`

export const Description = styled.p`
    font-size: 1.25rem;

    position: absolute;
    top: 10px;
    right: 0;
`

export const List = styled.ul`
    display: flex;
    column-gap: 26px;
    row-gap: 34px;
    flex-wrap: wrap;

    list-style: none;
`