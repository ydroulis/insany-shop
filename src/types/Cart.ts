
export type CartProduct = {
    id: number
    name: string
    description: string
    price: number
    image: string
    stock: number
    items?: number
}

export type Cart = {
    content: CartProduct[]
    total: number
}