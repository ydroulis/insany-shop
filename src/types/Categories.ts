export type Categories = Category[];

export type Category = {
    id: string,
    name: string,
    description: string,
    icon: string,
    productCount: number
};