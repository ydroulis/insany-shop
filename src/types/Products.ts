export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}