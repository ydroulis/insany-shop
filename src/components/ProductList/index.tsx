import React from 'react';

import * as S from './styles';
import { Products } from '@/types/Products';
import ProductCard from '../ProductCard';

interface ProductListProps {
    products: Products
    pageId?: string
}

const categories = [
    {
        id: "eletronicos",
        name: "Eletr√¥nicos",
        description: "Smartphones, laptops, consoles e mais",
        icon: "üì±",
        productCount: 4
    },
    {
        id: "roupas",
        name: "Roupas e Cal√ßados",
        description: "Moda masculina, feminina e acess√≥rios",
        icon: "üëï",
        productCount: 4
    },
    {
        id: "casa",
        name: "Casa e Decora√ß√£o",
        description: "M√≥veis, decora√ß√£o e utilidades dom√©sticas",
        icon: "üè†",
        productCount: 3
    },
    {
        id: "livros",
        name: "Livros",
        description: "Literatura, t√©cnicos, educacionais e mais",
        icon: "üìö",
        productCount: 3
    },
    {
        id: "esportes",
        name: "Esportes e Lazer",
        description: "Equipamentos esportivos e atividades ao ar livre",
        icon: "‚öΩ",
        productCount: 3
    }
]

const ProductList: React.FC<ProductListProps> = ({ products, pageId }) => {
    const hasProducts = products && products.length > 0;
    const category = categories.find(category => category.id === pageId)

    const sectionTitleId = 'product-list-title';
    const descriptionId = pageId ? 'product-list-description' : undefined;

    return (
        <S.Container
            aria-labelledby={sectionTitleId}
            aria-describedby={descriptionId}
            role="region"
        >
            <S.Title id={sectionTitleId}>{pageId ? category?.name : "Todos os Produtos"}</S.Title>
            {pageId && (
                <S.Description id={descriptionId}>{category?.description}</S.Description>
            )}

            {hasProducts ? (
                <S.List role="list">
                    {products.map((product) => (
                        <li key={product.id} role="listitem">
                            <ProductCard {...product} />
                        </li>
                    ))}
                </S.List>
            ) : (
                <p role="status" aria-live="polite">
                    Nenhum produto dispon√≠vel no momento.
                </p>
            )}
        </S.Container>
    );
}

export default ProductList;