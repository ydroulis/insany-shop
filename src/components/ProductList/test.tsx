import React from "react";
import { render, screen } from "@testing-library/react";
import ProductList from "./";
import { Products } from "@/types/Products";

jest.mock("../ProductCard", () => {
    return function MockProductCard({
        name,
        description,
    }: {
        name: string;
        description: string;
    }) {
        return (
            <div data-testid="mock-product-card">
                <h3>{name}</h3>
                <p>{description}</p>
            </div>
        );
    };
});

describe("<ProductList />", () => {
    const mockProducts: Products = [
        {
            id: 1,
            name: "Produto A",
            description: "Descrição do Produto A",
            price: 100,
            image: "/produto-a.png",
            category: "eletronicos",
            stock: 5,
            rating: 4.5,
        },
        {
            id: 2,
            name: "Produto B",
            description: "Descrição do Produto B",
            price: 200,
            image: "/produto-b.png",
            category: "eletronicos",
            stock: 2,
            rating: 3.8,
        },
    ];

    it("should render section with role='region' and associated title", () => {
        render(<ProductList products={mockProducts} />);

        const region = screen.getByRole("region", {
            name: /todos os produtos/i,
        });
        expect(region).toBeInTheDocument();

        const heading = screen.getByRole("heading", { name: /todos os produtos/i });
        expect(heading).toBeInTheDocument();
        expect(heading).toHaveAttribute("id", "product-list-title");
    });

    it("should render list of products when not empty", () => {
        render(<ProductList products={mockProducts} />);

        const list = screen.getByRole("list");
        expect(list).toBeInTheDocument();

        const items = screen.getAllByRole("listitem");
        expect(items).toHaveLength(mockProducts.length);

        expect(screen.getByText("Produto A")).toBeInTheDocument();
        expect(screen.getByText("Produto B")).toBeInTheDocument();
    });

    it("should not render fallback message when products are present", () => {
        render(<ProductList products={mockProducts} />);
        expect(
            screen.queryByText(/nenhum produto disponível/i)
        ).not.toBeInTheDocument();
    });

    it("should render acessible fallback message when products are empty", () => {
        render(<ProductList products={[]} />);

        const fallback = screen.getByRole("status");
        expect(fallback).toBeInTheDocument();
        expect(fallback).toHaveAttribute("aria-live", "polite");
        expect(fallback).toHaveTextContent(/nenhum produto disponível/i);
    });

    it("should render with pageId and show category name and description", () => {
        render(<ProductList products={mockProducts} pageId="eletronicos" />);

        const region = screen.getByRole("region", { name: /eletrônicos/i });
        expect(region).toBeInTheDocument();

        const heading = screen.getByRole("heading", { name: /eletrônicos/i });
        expect(heading).toHaveAttribute("id", "product-list-title");

        const description = screen.getByText(/smartphones, laptops, consoles e mais/i);
        expect(description).toBeInTheDocument();
        expect(description).toHaveAttribute("id", "product-list-description");

        expect(region).toHaveAttribute("aria-describedby", "product-list-description");
    });

    it("should render product cards with correct props passed", () => {
        render(<ProductList products={mockProducts} />);

        const cards = screen.getAllByTestId("mock-product-card");
        expect(cards).toHaveLength(mockProducts.length);

        expect(screen.getByText("Descrição do Produto A")).toBeInTheDocument();
        expect(screen.getByText("Descrição do Produto B")).toBeInTheDocument();
    });

    it("should match snapshot without pageId", () => {
        const { container } = render(<ProductList products={mockProducts} />);
        expect(container).toMatchSnapshot();
    });

    it("should match snapshot with pageId", () => {
        const { container } = render(
            <ProductList products={mockProducts} pageId="eletronicos" />
        );
        expect(container).toMatchSnapshot();
    });
});
