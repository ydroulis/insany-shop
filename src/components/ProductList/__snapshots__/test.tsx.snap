// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<ProductList /> should match snapshot with pageId 1`] = `
<div>
  <section
    aria-describedby="product-list-description"
    aria-labelledby="product-list-title"
    class="sc-bRKDuR bdSZYh"
    role="region"
  >
    <h2
      class="sc-hvigdm dpWcbF"
      id="product-list-title"
    >
      Eletrônicos
    </h2>
    <p
      class="sc-fhHczv hiFLoZ"
      id="product-list-description"
    >
      Smartphones, laptops, consoles e mais
    </p>
    <ul
      class="sc-ggWZvA eYZMRa"
      role="list"
    >
      <li
        role="listitem"
      >
        <div
          data-testid="mock-product-card"
        >
          <h3>
            Produto A
          </h3>
          <p>
            Descrição do Produto A
          </p>
        </div>
      </li>
      <li
        role="listitem"
      >
        <div
          data-testid="mock-product-card"
        >
          <h3>
            Produto B
          </h3>
          <p>
            Descrição do Produto B
          </p>
        </div>
      </li>
    </ul>
  </section>
</div>
`;

exports[`<ProductList /> should match snapshot without pageId 1`] = `
<div>
  <section
    aria-labelledby="product-list-title"
    class="sc-bRKDuR bdSZYh"
    role="region"
  >
    <h2
      class="sc-hvigdm dpWcbF"
      id="product-list-title"
    >
      Todos os Produtos
    </h2>
    <ul
      class="sc-ggWZvA eYZMRa"
      role="list"
    >
      <li
        role="listitem"
      >
        <div
          data-testid="mock-product-card"
        >
          <h3>
            Produto A
          </h3>
          <p>
            Descrição do Produto A
          </p>
        </div>
      </li>
      <li
        role="listitem"
      >
        <div
          data-testid="mock-product-card"
        >
          <h3>
            Produto B
          </h3>
          <p>
            Descrição do Produto B
          </p>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
